File: GuardianX-main/guardianx-frontend/src/types/finding.ts
export type FindingSeverity = "CRITICAL" | "HIGH" | "MEDIUM" | "LOW" | "UNKNOWN";

export type FindingStatus =
  | "OPEN"
  | "IN_PROGRESS"
  | "RESOLVED"
  | "FALSE_POSITIVE"
  | "ACCEPTED_RISK";

export interface Finding {
  id: number;
  title: string;
  severity: FindingSeverity;
  cve: string | null;
  cvss: number | null;
  status: FindingStatus;
  assigned_to: number | null;
  assigned_to_name: string | null;
  due_date: string | null;
  created_at: string;
  updated_at: string;
  asset_name: string | null;
  affected_service: string | null;
}

export interface FindingDetail {
  id: number;
  title: string;
  description: string | null;
  severity: FindingSeverity;
  cve: string | null;
  cvss: number | null;
  affected_asset: string | null;
  affected_service: string | null;
  recommendation: string | null;
  status: FindingStatus;
  assigned_to: number | null;
  assigned_to_name: string | null;
  notes: string | null;
  due_date: string | null;
  created_at: string;
  updated_at: string;
  provider: string | null;
  lookup_method: string | null;
  confidence: string | null;
}

export interface VulnerabilityIntelligence {
  id: string;
  cvss_score: number | null;
  published: string | null;
  last_modified: string | null;
  cwes: string[];
  affected_vendors: string[];
  affected_products: string[];
  epss_score: number | null;
  epss_percentile: number | null;
  exploited: boolean;
  exploit_status: string;
  guardianx_risk_score: number;
  threat_level: string;
  ai_summary: string;
  references: Array<{ url: string; source: string; tags: string[] }>;
  attack_techniques: Array<{ id: string; name: string; tactics: string[]; description: string }>;
}

export interface FindingIntelligenceResponse {
  status: "not_available" | "pending" | "ready" | "unavailable";
  intelligence: VulnerabilityIntelligence | null;
}

export interface FindingActivity {
  id: number;
  finding_id: number;
  user_id: number | null;
  username: string | null;
  action: string;
  old_value: string | null;
  new_value: string | null;
  created_at: string;
}

export interface FindingStats {
  total: number;
  open: number;
  in_progress: number;
  resolved: number;
  false_positive: number;
  accepted_risk: number;
  by_severity: Record<string, number>;
}

export interface Assignee {
  id: number;
  username: string;
}

export interface FindingTriageUpdate {
  status?: FindingStatus;
  assignee_id?: number | null;
  notes?: string | null;
  due_date?: string | null;
}

export interface BulkStatusUpdate {
  ids: number[];
  status: FindingStatus;
}

export interface BulkUpdateResult {
  updated: number;
  ids: number[];
}

export interface FindingListResponse {
  items: Finding[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface FindingQueryParams {
  page?: number;
  size?: number;
  severity?: FindingSeverity;
  status?: FindingStatus;
  asset?: string;
  cve?: string;
  search?: string;
  assigned?: "me" | "unassigned";
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

File: GuardianX-main/backend/app/schemas/finding.py
from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field


class FindingStatus(str, Enum):
    OPEN = "OPEN"
    IN_PROGRESS = "IN_PROGRESS"
    RESOLVED = "RESOLVED"
    FALSE_POSITIVE = "FALSE_POSITIVE"
    ACCEPTED_RISK = "ACCEPTED_RISK"


class FindingListItem(BaseModel):
    id: int
    title: str
    severity: str
    cve: str | None = None
    cvss: float | None = None
    status: FindingStatus
    assigned_to: int | None = None
    assigned_to_name: str | None = None
    due_date: datetime | None = None
    created_at: datetime
    updated_at: datetime
    asset_name: str | None = None
    affected_service: str | None = None

    model_config = {
        "from_attributes": True,
    }


class FindingListResponse(BaseModel):
    items: list[FindingListItem] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    size: int = 20
    pages: int = 0


class FindingDetailResponse(BaseModel):
    id: int
    title: str
    description: str | None = None
    severity: str
    cve: str | None = None
    cvss: float | None = None
    affected_asset: str | None = None
    affected_service: str | None = None
    recommendation: str | None = None
    status: FindingStatus
    assigned_to: int | None = None
    assigned_to_name: str | None = None
    notes: str | None = None
    due_date: datetime | None = None
    created_at: datetime
    updated_at: datetime
    provider: str | None = None
    lookup_method: str | None = None
    confidence: str | None = None


class FindingIntelligenceResponse(BaseModel):
    """Asynchronous enrichment state for a CVE-backed finding."""

    status: Literal["not_available", "pending", "ready", "unavailable"]
    intelligence: dict | None = None


class FindingStatusUpdate(BaseModel):
    status: FindingStatus


class FindingTriageUpdate(BaseModel):
    status: FindingStatus | None = None
    assignee_id: int | None = None
    notes: str | None = None
    due_date: datetime | None = None


class FindingActivityResponse(BaseModel):
    id: int
    finding_id: int
    user_id: int | None = None
    username: str | None = None
    action: str
    old_value: str | None = None
    new_value: str | None = None
    created_at: datetime


class BulkFindingsStatusUpdate(BaseModel):
    ids: list[int] = Field(min_length=1, max_length=500)
    status: FindingStatus


class BulkUpdateResponse(BaseModel):
    updated: int
    ids: list[int] = Field(default_factory=list)


class FindingStatsResponse(BaseModel):
    total: int
    open: int
    in_progress: int
    resolved: int
    false_positive: int
    accepted_risk: int
    by_severity: dict[str, int] = Field(default_factory=dict)


class AssigneeResponse(BaseModel):
    id: int
    username: str
